Custom feedback messages

paulballesty opened the pull request:
I need to use the library in an internationalized website, which means that I need the feedback messages in other languages. I added the optional third parameter `options`, in where you can pass a `feedback_messages` Array-like object with custom messages.

Do you guys think it worth adding this to the library?
#### Example usages:
##### Redifining all the messages

``` js
zxcvbn('example', ['paul'], {feedback_messages: ['Utilice varias palabras, evite frases comunes', ...]});
```
##### Changing an specific message

``` js
zxcvbn('example', ['paul'], {feedback_messages: {9: 'Sequences are easy to guess'}});
```

paulballesty committed:
custom feedback messages

peterver commented:
+1 

definitely an upide to this merge as 2 issues are open regarding this ( #118 and #126 ), can someone take a look at why the travis ci builds are failing ? :]

9and3r commented:
+1

This would be great to use it with other languges

thedotedge commented:
👍

wje97951 commented:
+1

paulballesty commented:
@lowe any interest on this? I am getting emails from people asking me how to use it. Based on the +1s in this PR, it seems users need a way to modify the feedback messages.

Happy to change the code if you don't like the approach.

antys10g commented:
+1

ck-ws commented:
+1

lowe commented:
Hi @paulballesty,

Thanks for the PR. I'm happy to merge but have two requests to work through first: 

* Can you add short documentation in the README that includes a brief example, under the "Usage" section (before "Performance")?

* Instead of an array object, what do you think about using an object? I think that will be more self-documenting and also make it easier to add messages in the middle of the list. Something like: 

```
messages =
  use_a_few_words: "Use a few words, avoid common phrases"
  no_need_for_mixed_chars: "No need for symbols, digits, or uppercase letters"
  ...

# instead of -- 
# extra_feedback = @get_message(2) # Add another word or two. Uncommon words are better.
extra_feedback = messages.use_few_words # no need to copy/paste the message, which is costlier to maintain as messages change. 
```

* Lastly, I would suggest that if a key is missing from the optional `messages` object, it should result in an absence of that message in the corresponding `suggestions` array. That way, if a user doesn't like a particular feedback string, they can exclude it in their rewording/translation.

cdroulers commented:
I agree with @lowe for the dictionary, otherwise numbers aren't very descriptive when overriding!

paulballesty committed:
feedback (not tested yet)

the-question commented:
@paulballesty's PR (now with a dictionnary thanks to 55b9ce5) looks nice and is a must have! Any chance having it merged soon @lowe?
It's been more than a year without any news.

paulballesty commented:
@the-question I didn't have time to test the last changes properly. I'll try to find some time this week. Sorry for the delay!

lpavlicek commented:
Thank @paulballesty for PR. I tried to use it but I encountered a few problems:

-  default value for options in zxcvbn function: 
`zxcvbn = (password, user_inputs = [], options = {})`
Empty object means empty feedback or default feedback ? I suggest` {} `for empty feedback and `null `for default feedback, so I recommend change function header to:
`zxcvbn = (password, user_inputs = [], options = null)`

- I'm beginner in Javscript/Coffeescript and I don't understand next construction from main function:
`{ feedback_messages } = options`
In my code I change it to 
`feedback_messages = options`

- If I disable some suggestion then feedback contains empty string. Next code:
```
feedback.get_feedback(0, '', {
      'no_need_for_mixed_chars' : 'modified suggestion',
      'use_a_few_words' : null
      }),
```
returns empty string in suggestions unnecessary:
```
 {
      suggestions: [ '', 'modified suggestion' ],
      warning: ''
    }
```
so in function `build_feedback` i changed
`suggestions.push message if message?`
to
`suggestions.push message if message? and message != ''`

I attached feedback.patch and also tape tests for feedback (test/test-feedback.coffee).

[zxcvbn_pr_patch.zip](https://github.com/dropbox/zxcvbn/files/909796/zxcvbn_pr_patch.zip)

paulballesty committed:
feedback test with default messages

paulballesty committed:
unused layout var & console.log removal

paulballesty committed:
Changes strategy to skip messages and add tests for custom messages

paulballesty committed:
Moves user_inputs attribute inside options

paulballesty committed:
update docs

paulballesty commented:
@lowe finally found time to work on this 😄 (public holiday in Australia 🇦🇺 ).

> Can you add short documentation in the README that includes a brief example, under the "Usage" section (before "Performance")?

In the last commits I moved the `user_inputs` parameter inside an `options` parameter (I made it backward compatible checking the type of the object passed by the consumer). I did this because I wouldn't like to force the user to pass something on `user_inputs` when they just want to use `feedback_messages`. I think this could be helpful if there's any intention of adding another optional parameter in the future. LMK your opinion on this.

> Instead of an array object, what do you think about using an object?

Done

> Lastly, I would suggest that if a key is missing from the optional messages object, it should result in an absence of that message in the corresponding suggestions array. That way, if a user doesn't like a particular feedback string, they can exclude it in their rewording/translation. When a code path would have added that feedback string, it'll instead do nothing if it can't find the key.

I did something a bit different here. Instead of excluding the message if it's not present in `feedback_messages`, I am only excluding it if the key is present but the value is *falsy* (`null`, `undefined`, `false`, etc) . If the key is missing, then we default to the in-app message. The argument for this decision is that I think it would be much more common to want to delete or modify a specific message, and we would be forcing users to copy and paste all the other messages that they don't want to modify.

I included [tests](https://github.com/dropbox/zxcvbn/pull/124/files#diff-3ca125bfc62c24f9841fee0c79bcb37c) for the feedback component, I think that's gonna make things easier in the future 😄 . Also cherry-picked those tests in another branch without all the custom feedback logic to check that these changes aren't breaking anything in the feedback logic, you can check that [here](https://github.com/paulballesty/zxcvbn/tree/before-custom-feedback).

paulballesty commented on src/feedback.coffee:
```diff
@@ -1,120 +1,144 @@
         @get_dictionary_match_feedback match, is_sole_match
 
       when 'spatial'
-        layout = match.graph.toUpperCase()
```
Was this used somewhere?

paulballesty commented on src/main.coffee:
```diff
@@ -5,7 +5,15 @@ feedback = require './feedback'
+    { user_inputs, feedback_messages } = options
+  else
+    user_inputs = []
+    feedback_messages = {}
```
^ this is the code moving `user_inputs` in a `options` object

lpavlicek replied:
`options` are a bit better because they make future changes easier. I get error when compile, so I suggest the following version:

```
zxcvbn = (password, options = {}) ->
  user_inputs = []
  feedback_messages = {}
  if options instanceof Array
    user_inputs = options # backward-compatibility
  else
    user_inputs = options.user_inputs if options.user_inputs
    feedback_messages = options.feedback_messages if options.feedback_messages
```

lpavlicek commented on src/feedback.coffee:
```diff
@@ -1,120 +1,144 @@
+    suggestions = []
+    for suggestion_key in suggestion_keys
+      message = @get_message(suggestion_key)
+      suggestions.push message if message?
```
Please add message to suggestions only if not empty:
`suggestions.push message if message? and message != ''`

Array of suggestions with empty string like `[ '' ]` or `[ '', 'Avoid sequences']` is complication for displaying feedback. Empty string occurs if message is present in custom messages and has a falsy value.

clarkewing commented:
Hi @lowe, any chance of merging this in the near future? Seems ready to me!

olegomon commented on src/main.coffee:
```diff
@@ -5,7 +5,15 @@ feedback = require './feedback'
+zxcvbn = (password, options = {}) ->
+  if options instanceof Array
+    user_inputs = options # backward-compatibility
+  else if typeof options === 'object'
```
@paulballesty I don't know CoffeeScript well but strict equality checks using "===" is  not valid syntax in CoffeeScript my local build failed using your PR. CoffeeScript converts all loose checks into strict equality checks by default so using "===" is not needed here.

olegomon commented on src/main.coffee:
```diff
@@ -5,7 +5,15 @@ feedback = require './feedback'
+  if options instanceof Array
+    user_inputs = options # backward-compatibility
+  else if typeof options === 'object'
+    { user_inputs, feedback_messages } = options
```
@paulballesty In that case `user_input` should be initialised with an empty array to mimic the old behaviour. Otherwise an exception is thrown in [main.coffee L20](https://github.com/paulballesty/zxcvbn/blob/ace4b01a3215350f316f21e25f03e7cdbbee6f12/src/main.coffee#L20) when you trying to iterate over the values of an undefined user_input array.

olegomon committed:
- removed strict equality check for coffee script
- added default value for user_inputs if options is not defined
- added default value for feedback_messages if options is not defined
- added dist build

Sublime1 commented:
Hi Paul
I'd love to try out your localisation changes. Any change you'd generate the .js dist file so I can try it out?

thanks

ctavan commented:
@paulballesty @olegomon I would find this PR extremely helpful for my use cases. What holds us back from getting it merged? Is there anything I can do to help getting this merged?

olegomon commented:
@ctavan I don't know why this PR didn't get enough attention yet. So I don't know what we can do to get this merged.

@Sublime1 There is one which you can get from my fork.

Sublime1 commented:
Thanks @olegomon but I can't find it. Would you mind posting a link? I'm specifically looking for a .js dist file that I can just drop in.

olegomon commented:
@Sublime1 here is the repository https://github.com/prospective/zxcvbn where you’ll find the .js file. 

@paulballesty could you please review my PR? If you accept it we may have a better chance to get it merged in the main repository. https://github.com/paulballesty/zxcvbn/pull/1

paulballesty committed:
Merge pull request #1 from prospective/master

ghost commented:
Hi, any chance to have t his merged any time soon?

wwelles commented:
Also would love to see this in master.

ctavan commented:
In case anyone is interested, we have temporarily released this branch as npm package: [`@contentpass/zxcvbn`](https://www.npmjs.com/package/@contentpass/zxcvbn)

Pierre-Sassoulas commented:
[Here's a MIT licensed french translation of most of the feedback strings](https://github.com/Pierre-Sassoulas/django-zxcvbn-password-validator/blob/master/django_zxcvbn_password_validator/locale/fr/LC_MESSAGES/django.po). It's a .po file for another project, but you can probably adapt it quickly.

revington commented:
Please merge this PR as soon as possible! the world is full of non english speakers who need secure passwords! :)

jeroencornelissen commented:
+1

tracker1 commented:
NOTE: given that this changes the interface, it should be published as a major version bump, so it doesn't interfere with exiting users of yarn/npm.

tracker1 commented on README.md:
```diff
@@ -203,7 +203,28 @@ result.calc_time  # how long it took zxcvbn to calculate an answer,
 
-The optional `user_inputs` argument is an array of strings that zxcvbn will treat as an extra dictionary. This can be whatever list of strings you like, but is meant for user inputs from other fields of the form, like name and email. That way a password that includes a user's personal information can be heavily penalized. This list is also good for site-specific vocabulary — Acme Brick Co. might  ...
+The optional `options` argument is an object that can contain the following optional properties:
+- `user_inputs` is an array of strings that zxcvbn will treat as an extra dictionary. This can be whatever list of strings you like, but is meant for user inputs from other fields of the form, like name and email. That way a password that includes a user's personal information can be heavily penalized. This list is also good for site-specific vocabulary — Acme Brick Co. might want to include ['ac ...
```
should probably be `userInputs` and `feedbackMessages` respectively... this is JavaScript.

tracker1 commented on src/main.coffee:
```diff
@@ -5,7 +5,15 @@ feedback = require './feedback'
-zxcvbn = (password, user_inputs = []) ->
+zxcvbn = (password, options = {}) ->
+  if options instanceof Array
+    user_inputs = options # backward-compatibility
```
`userInputs` and `feedbackMessages` ?

paulballesty commented:
Closing this PR. It's been open for more than 2 years.

L3o-pold commented:
This need to be merged not closed!

cdroulers commented:
@tracker1 Looks like it was backwards compatible with line 9 of `src/main.coffee`, no?

tracker1 commented:
@cdroulers It does seem that way... the issue, to me is the prior version didn't force a convention, as it was an argument.  Now that you're passing an object, imho the object structure should conform to JS naming conventions (camel-case as opposed to snake-case).

cdroulers commented:
@tracker1 Oh, absolutely, I posted this in reference to your previous comment!

xtrasimplicity commented:
@paulballesty, are you open to this sort of feature being merged (whether in this form or another)? I understand wanting to tidy up a bit, but it looked like it was starting to gain some traction again?

paulballesty commented:
@xtrasimplicity I am not an owner of the repo so I cannot merge the feature myself. I left the PR open for more than 2 years. I closed it assuming there's no intention of merging it.

xtrasimplicity commented:
Thanks for clarifying, @paulballesty - much appreciated. :)

@lowe, Are you able to shed some light on this?

kasperwandahl commented:
@lowe @paulballesty Just found this PR, and this is something I need :+1: Can I do anything to help test, tidy up, etc? Would be nice to have it merged and bumped to a new version.

Aggror commented:
Translation options would be a very common thing in my opinion. I thought I found a decent library, but seeing a PR like this not being used to its potential is a bit of a let down.

jesobreira commented:
Is this going somewhere?

For now I'm using [this package](https://www.npmjs.com/package/@contentpass/zxcvbn) published by @ctavan (thanks!).

bernardsaucier commented:
Yet another one of those "looks good, looks good, looks good, oh crap, need to find another library"...

Saggre commented:
I need this feature and I'd be ready to rebase this onto the current master. @lowe would you check and merge the resulting pr?
